### PR TITLE
[WiP] Pathfinding rework

### PR DESCRIPTION
**Pull request by Coolthulhu**
Big pathfinding rework that is supposed to make pathfinding fast enough to use for zombie hordes.

Will work by inverting the pathfinding direction: instead of searching from own position to target position, it will search from target to searcher. This way we can re-use same pathfinding data to find more than one target with minimal performance penalty.

A disadvantage of this approach is that topology data can't be used - A\* has to be "downgraded" to Dijkstra. Algorithm will still use A\* if only a small number (2? 4?) of searches would be made, because we already have complete A\* written and A\* can be more than 3 times faster in some cases (typically only in large search spaces, though).

Monster planning phase will have to be altered. Monsters won't find paths in planning phase, but after it. They will register which targets they want and only once all monsters finish the planning phase, they will receive the plans.
This means monsters would have to go through their planning->action phases in entire factions. For example, all zombies execute their turn first, then all animals etc. This isn't a hard requirement - simply a good way to handle it without causing recalculations of paths. Recalculating paths wouldn't necessarily be expensive enough to get slow, not even if it was done several times per turn.

As a result of the above (if strict per-faction planning was implemented), zombies would exhibit a pileup behavior, where they'd gang on creatures for a whole turn even if it died in the first hit and would also have "inertia" where they'd blunder into space where an enemy was before a hulk knocked it away.
I don't think this would be a bad thing - most players wouldn't notice, and for those that would, it could be considered a feature.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Results look promising. The slowdown was very manageable with low number of targets (ie. just the player and 2 wild animals). If zombie pathfinding was limited to orders from leaders, it could easily be handled here.

Got zeds to surround properly too. Even surround + pile up behind the surrounding ones.

**Comment by kevingranade** [comment]
Why do you want zombies pathing again?  It makes sense for some other
monsters, but zombies really should be dumb as rocks.

**Comment by Coolthulhu** [comment]
Zombies are mostly for stress-testing - there's a lot of those in game, with different stats.
Plus, it could be an ability of zombie masters/necros - marking a player, ordering all the zeds to path instead of charging.

**Comment by kevingranade** [comment]
In that case please make it clear you don't actually mean for the zombies to do advanced pathing.
When a leader gives them directions, it should happen on the leader's action, not when the monster acts.  This is how we handle brain blobs and the thriller.

That having been said, you now don't have a stated purpose for the PR, what features are you enabling here?  "A manageable amount of slowdown" sounds VERY un-exciting to me, I've done a crapton of work making various parts of the game acceptably fast, and speed regressions need to be justified with very cool features, and I'm not hearing about any here.

**Comment by Coolthulhu** [comment]
The most basic idea was to eliminate two zombie behaviors that don't really make sense.
First one is piling up in rough triangles and being "kiteable" because of that - currently zombies are incapable of surrounding any semi-competent player.
By giving them short ranged pathing (even as much as 4 tiles), they'd surround the player when unable to attack rather than hoping to stumble in proper direction.
Additionally when charging in a group, they wouldn't stumble when they pick a square that contains an ally (this makes zombie hordes impossible - they quickly separate into a sparse group). Though this one could be also partially fixed just by making them use `wander_next`-like mechanics when picking a tile with an ally on it.
Piling up on each other makes sense when we think of zombies as symbols on sparse square grid, but loses this sense when we try to imagine how the situation looks from character's perspective.
A small adjustment of pathing could result in them surrounding the player even when attacking, representing the zombies behind pushing each other.

Pathing would also have an advantage of letting zeds not get locked up trying to bash down cars when chasing the player. Once again this makes sense from a square point of view, but "IRL" even insect intelligence is enough to get around obstacles as obvious as a car.

During the testing I used advanced pathing (no distance limit), because makes bugs obvious, while simplified one (low max. distance) could easily mask them.

As for slowdown - this was mostly just a stress test. If the slowdown was bad, I'd have to either drop the whole idea or invent some clever workaround. For a small number of targets (just the player), it would be unnoticeable, especially when compared to slowdown caused by a great number of zombies just existing.

To sum up: making zombies harder to kite and shake off by leaving them in cars without slowing the game down.

**Comment by KA101** [comment]
My take is that zeds are common enough that I'd rather folks be able to do some amount of kiting.  (New folks should be able to succeed at something.)  We have nastier and faster zeds to defeat kiters, as well as ranged ones.  Global benefits to their AI should be strictly in the presence of a master or brain blob.

**Comment by Coolthulhu** [comment]
It wouldn't completely eliminate kiting, just make it more resistant to player "dancing".
An observant player can easily take the entire horde 1-2 zeds at a time. Makes horde only as scary as the number of special zombies in it.

**Comment by kevingranade** [comment]
Tl;dr no pathing for zombies

While I agree the issues you're trying to address are valid, I don't like
making zombies 'smarter' as fixes.  With surrounding the player, how about
making them actually shove each other if they get in the way instead of
making them path and justifying it as them shoving each other?  In
particular I've been meaning to make them shove each other through windows
when possible for a while now.
As for pathing around cars, that is precisely the kind of thinking ahead
they should not be doing.  I could see have them 'give up' on bashing a
square and trying something different, but anticipating not being able to
bash something or deciding it would be faster to go around instead is too
much thinking.

**Comment by Coolthulhu** [comment]
I don't like the idea of zombies being too dumb to path around cars - makes them literally dumber than insects. Plus, it will really show if we ever get DF-style z-level trees and similar structures (climb a tree, move onto branch far from trunk = zombieproof).

Closing for now - no other swarmers would really benefit from it, at least for now when most of the terrain (anything but cities or special structures) is rather featureless. Mi-gos, tankbots etc. are rare enough that if they'd path, they'd use the regular A* anyway.

**Comment by kevingranade** [comment]
You generally should be zombieproof in a tree, it's just that they'll never
leave the area.

**Comment by Coolthulhu** [comment]
Spears having reach is not just a suggestion, but a bounty issue.
Speartree may be the new windowknife.

**Comment by kevingranade** [comment]
Could be, make sure to check that stuck chance though ;)

**Comment by KA101** [comment]
>You generally should be zombieproof in a tree

Until the hulk, Jabberwock, hunter, or spitter come by and knock the tree down/climb it/spit at you directly.  ;-)

**Comment by Coolthulhu** [comment]
Hunter and spitter yeah, but hulk and jabberwock would depend on player leading them onto the tree trunk.
They'd most likely just stand directly under the player and kill all other zeds around.

**Comment by Rivet-the-Zombie** [comment]
Or knock the tree down.

**Comment by kevingranade** [comment]
The player is on a branch several tiles away from the tree trunk (This is
assuming a lot of changes about the game), so the hulks wouldn't target the
trunk directly, so the only way that would happen would be if they happened
to wander into the trunk and smash it.
This is very much how I would want it to work, so from my point of view
it's an argument against giving them pathing (unless a necro or master
shows up, in which case all the zombies start smashing the tree).
